### ISSUE TITLE
what is predict_q indicate?

### ISSUE THREAD

**Issue by longshot-pting** [zhenyuwei2003/DRO-Grasp#2]
dear author, I wonder what is predict_q indicate? In addition, how can I get 6-DOF pose and how can I visualize grasping process in IasscGYM which I tried set '--use_gui = True' when runned example_isaac.py did not work.
I will appreciate if you help!

**Comment by zhenyuwei2003** [comment]
`predict_q` represents the optimization result with a shape of `(B, 6 + DOF)`, corresponding to the grasp joint values output by the entire pipeline. Since we have modified the hand URDF to include dummy joints representing the 6D pose, the first six values of `predict_q` correspond to the 6D pose of the robotic hand.

Regarding the visualization of the grasping process, could you provide more details about the issues you encountered when running `example_isaac.py`? I have tested it myself, and the visualization works properly when uncommenting [this line](https://github.com/zhenyuwei2003/DRO-Grasp/blob/d740c0955851c6a07563f2796efd434ed7480edc/validation/validate_utils.py#L88). If you are running the code on a server without a graphical interface, you might need to address additional configuration issues to enable visualization.

**Comment by longshot-pting** [comment]
Thanks for answering!
In validate_utils.py I uncomment the '--use_gui', and in isaac_main.py I set 'use_gui: bool = True' in def isaac_main. However, I didn't get isaac_gym visualization when I runned example_isaac.py. Here is my code details：
validate_utils.py:
`def validate_isaac(robot_name, object_name, q_batch, gpu: int = 0):
    """
    Wrap function for subprocess call (isaac_main.py) to avoid Isaac Gym GPU memory leak problem.

    :param robot_name: str
    :param object_name: str
    :param q_batch: torch.Tensor, joint values to validate
    :param gpu: int
    :return: (list<bool>, list<float>), success list & info list
    """
    os.makedirs(os.path.join(ROOT_DIR, 'tmp'), exist_ok=True)
    q_file_path = str(os.path.join(ROOT_DIR, f'tmp/q_list_validate_{gpu}.pt'))
    torch.save(q_batch, q_file_path)
    batch_size = q_batch.shape[0]
    args = [
        'python',
        os.path.join(ROOT_DIR, 'validation/isaac_main.py'),
        '--mode', 'validation',
        '--robot_name', robot_name,
        '--object_name', object_name,
        '--batch_size', str(batch_size),
        '--q_file', q_file_path,
        '--gpu', str(gpu),
        # '--use_gui', 'True'
    ]`
isaac_main.py:
`import os
import sys
import json
import argparse
import warnings
from termcolor import cprint

ROOT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.append(ROOT_DIR)

from validation.isaac_validator import IsaacValidator  # IsaacGym must be imported before PyTorch
from utils.hand_model import create_hand_model
from utils.rotation import q_rot6d_to_q_euler

import torch


def isaac_main(
    mode: str,
    robot_name: str,
    object_name: str,
    batch_size: int,
    q_batch: torch.Tensor = None,
    gpu: int = 0,
    use_gui: bool = True
):
    """
    For filtering dataset and validating grasps.

    :param mode: str, 'filter' or 'validation'
    :param robot_name: str
    :param object_name: str
    :param batch_size: int, number of grasps in Isaac Gym simultaneously
    :param q_batch: torch.Tensor (validation only)
    :param gpu: int, specify the GPU device used by Isaac Gym
    :param use_gui: bool, whether to visualize Isaac Gym simulation process
    :return: success: (batch_size,), bool, whether each grasp is successful in Isaac Gym;
             q_isaac: (success_num, DOF), torch.float32, successful joint values after the grasp phase
    """
    if mode == 'filter' and batch_size == 0:  # special judge for num_per_object = 0 in dataset
        return 0, None
    if use_gui:  # for unknown reason otherwise will segmentation fault :(
        gpu = 0

    data_urdf_path = os.path.join(ROOT_DIR, 'data/data_urdf')
    urdf_assets_meta = json.load(open(os.path.join(data_urdf_path, 'robot/urdf_assets_meta.json')))
    robot_urdf_path = urdf_assets_meta['urdf_path'][robot_name]
    object_name_split = object_name.split('+') if object_name is not None else None
    # object_urdf_path = f'{object_name_split[0]}/{object_name_split[1]}/{object_name_split[1]}.urdf'
    object_urdf_path = f'{object_name_split[0]}/{object_name_split[1]}/coacd_decomposed_object_one_link.urdf'

    hand = create_hand_model(robot_name)
    joint_orders = hand.get_joint_orders()

    simulator = IsaacValidator(
        robot_name=robot_name, 
        joint_orders=joint_orders, 
        batch_size=batch_size,
        gpu=gpu, 
        is_filter=(mode == 'filter'),
        use_gui=use_gui
    )
    print("[Isaac] IsaacValidator is created.")

    simulator.set_asset(
        robot_path=os.path.join(data_urdf_path, 'robot'),
        robot_file=robot_urdf_path[21:],  # ignore 'data/data_urdf/robot/'
        object_path=os.path.join(data_urdf_path, 'object'),
        object_file=object_urdf_path
    )
    simulator.create_envs()
    print("[Isaac] IsaacValidator preparation is done.")

    if mode == 'filter':
        dataset_path = os.path.join(ROOT_DIR, f'data/CMapDataset/cmap_dataset.pt')
        metadata = torch.load(dataset_path)['metadata']
        q_batch = [m[1] for m in metadata if m[2] == object_name and m[3] == robot_name]
        q_batch = torch.stack(q_batch, dim=0).to(torch.device('cpu'))
    if q_batch.shape[-1] != len(joint_orders):
        q_batch = q_rot6d_to_q_euler(q_batch)

    simulator.set_actor_pose_dof(q_batch.to(torch.device('cpu')))
    success, q_isaac = simulator.run_sim()
    simulator.destroy()

    return success, q_isaac


# for Python scripts subprocess call to avoid Isaac Gym GPU memory leak problem
if __name__ == '__main__':
    warnings.simplefilter(action='ignore', category=FutureWarning)
    parser = argparse.ArgumentParser()
    parser.add_argument('--mode', type=str, required=True)
    parser.add_argument('--robot_name', type=str, required=True)
    parser.add_argument('--object_name', type=str, required=True)
    parser.add_argument('--batch_size', type=int, required=True)
    parser.add_argument('--q_file', type=str)
    parser.add_argument('--gpu', default=0, type=int)
    parser.add_argument('--use_gui', action='store_true')
    args = parser.parse_args()

    print(f'GPU: {args.gpu}')
    assert args.mode in ['filter', 'validation'], f"Unknown mode: {args.mode}!"
    q_batch = torch.load(args.q_file, map_location=f'cpu') if args.q_file is not None else None
    success, q_isaac = isaac_main(
        mode=args.mode,
        robot_name=args.robot_name,
        object_name=args.object_name,
        batch_size=args.batch_size,
        q_batch=q_batch,
        gpu=args.gpu,
        use_gui=args.use_gui
    )

    success_num = success.sum().item()
    if args.mode == 'filter':
        print(f"<{args.robot_name}/{args.object_name}> before: {args.batch_size}, after: {success_num}")
        if success_num > 0:
            q_filtered = q_isaac[success]
            save_dir = str(os.path.join(ROOT_DIR, 'data/CMapDataset_filtered', args.robot_name))
            os.makedirs(save_dir, exist_ok=True)
            torch.save(q_filtered, os.path.join(save_dir, f'{args.object_name}_{success_num}.pt'))
    elif args.mode == 'validation':
        cprint(f"[{args.robot_name}/{args.object_name}] Result: {success_num}/{args.batch_size}", 'green')
        save_data = {
            'success': success,
            'q_isaac': q_isaac
        }
        os.makedirs(os.path.join(ROOT_DIR, 'tmp'), exist_ok=True)
        torch.save(save_data, os.path.join(ROOT_DIR, f'tmp/isaac_main_ret_{args.gpu}.pt'))
`

**Comment by longshot-pting** [comment]
> `predict_q` represents the optimization result with a shape of `(B, 6 + DOF)`, corresponding to the grasp joint values output by the entire pipeline. Since we have modified the hand URDF to include dummy joints representing the 6D pose, the first six values of `predict_q` correspond to the 6D pose of the robotic hand.
> 
> Regarding the visualization of the grasping process, could you provide more details about the issues you encountered when running `example_isaac.py`? I have tested it myself, and the visualization works properly when uncommenting [this line](https://github.com/zhenyuwei2003/DRO-Grasp/blob/d740c0955851c6a07563f2796efd434ed7480edc/validation/validate_utils.py#L88). If you are running the code on a server without a graphical interface, you might need to address additional configuration issues to enable visualization.

Thanks for answering!
In validate_utils.py I uncomment the '--use_gui', and in isaac_main.py I set 'use_gui: bool = True' in def isaac_main. However, I didn't get isaac_gym visualization when I runned example_isaac.py. Here is my code details：
validate_utils.py:
`    args = [
        'python',
        os.path.join(ROOT_DIR, 'validation/isaac_main.py'),
        '--mode', 'validation',
        '--robot_name', robot_name,
        '--object_name', object_name,
        '--batch_size', str(batch_size),
        '--q_file', q_file_path,
        '--gpu', str(gpu),
        # '--use_gui', 'True'
    ]`
isaac_main.py:
`def isaac_main(
    mode: str,
    robot_name: str,
    object_name: str,
    batch_size: int,
    q_batch: torch.Tensor = None,
    gpu: int = 0,
    use_gui: bool = True
):`

**Comment by zhenyuwei2003** [comment]
Instead of adding `'True'` after `'--use_gui'`, you should just uncomment `'--use_gui'`. So it should be `args = ['python', os.path.join(ROOT_DIR, 'validation/isaac_main.py'), '--mode', 'validation', '--robot_name', robot_name, '--object_name', object_name, '--batch_size', str(batch_size), '--q_file', q_file_path, '--gpu', str(gpu), '--use_gui']`. The args is used [here](https://github.com/zhenyuwei2003/DRO-Grasp/blob/d740c0955851c6a07563f2796efd434ed7480edc/validation/isaac_main.py#L99) (`store_true` means it will be set true when giving the args instead of manually set `True`). Modifying the default param of `isaac_main()` is useless since we already pass the param [here](https://github.com/zhenyuwei2003/DRO-Grasp/blob/d740c0955851c6a07563f2796efd434ed7480edc/validation/isaac_main.py#L112C15-L112C29). Hope it can solve your problem.

**Comment by longshot-pting** [comment]
> Instead of adding `'True'` after `'--use_gui'`, you should just uncomment `'--use_gui'`. So it should be `args = ['python', os.path.join(ROOT_DIR, 'validation/isaac_main.py'), '--mode', 'validation', '--robot_name', robot_name, '--object_name', object_name, '--batch_size', str(batch_size), '--q_file', q_file_path, '--gpu', str(gpu), '--use_gui']`. The args is used [here](https://github.com/zhenyuwei2003/DRO-Grasp/blob/d740c0955851c6a07563f2796efd434ed7480edc/validation/isaac_main.py#L99) (`store_true` means it will be set true when giving the args instead of manually set `True`). Modifying the default param of `isaac_main()` is useless since we already pass the param [here](https://github.com/zhenyuwei2003/DRO-Grasp/blob/d740c0955851c6a07563f2796efd434ed7480edc/validation/isaac_main.py#L112C15-L112C29). Hope it can solve your problem.

Thanks for replying! I have solved this problem!
